From 8a2ac08643199f2610109fe18bca66b59d492e6b Mon Sep 17 00:00:00 2001
From: xplodwild <xplod@ouverta.fr>
Date: Wed, 13 Nov 2013 00:04:51 +0100
Subject: [PATCH] libutils: Forward-port an old callstack method

Change-Id: I77a0dbc6214e662c2217f25cdf3f3883cd959bb3
---

diff --git a/include/utils/CallStack.h b/include/utils/CallStack.h
index 2056751..cae64a2 100644
--- a/include/utils/CallStack.h
+++ b/include/utils/CallStack.h
@@ -72,6 +72,13 @@
     void log(const char* logtag,
              android_LogPriority priority = ANDROID_LOG_DEBUG,
              const char* prefix = 0) const;
+#ifdef MTK_MT6577
+    // Dump a stack trace to the log
+    void dump(const char* prefix = 0) const;
+#endif
+
+    // Dump a stack trace to the log using the supplied logtag
+    void dump(const char* logtag, const char* prefix = 0) const;
 
     // Dump a stack trace to the specified file descriptor.
     void dump(int fd, int indent = 0, const char* prefix = 0) const;
diff --git a/libutils/Android.mk b/libutils/Android.mk
index 720443e..90a6931 100644
--- a/libutils/Android.mk
+++ b/libutils/Android.mk
@@ -106,6 +106,10 @@
 LOCAL_CFLAGS += -DALIGN_DOUBLE
 endif
 
+ifeq ($(TARGET_BOARD_PLATFORM),mt6577)
+LOCAL_CFLAGS += -DMTK_MT6577
+endif
+
 LOCAL_C_INCLUDES += \
 		bionic/libc/private \
 		external/zlib
diff --git a/libutils/CallStack.cpp b/libutils/CallStack.cpp
index 4ceaa7c..f911e13 100644
--- a/libutils/CallStack.cpp
+++ b/libutils/CallStack.cpp
@@ -114,6 +114,21 @@
     print(printer);
 }
 
+#ifdef MTK_MT6577
+void CallStack::dump(const char* prefix) const {
+    backtrace_symbol_t symbols[mCount];
+
+    get_backtrace_symbols(mStack, mCount, symbols);
+    for (size_t i = 0; i < mCount; i++) {
+        char line[MAX_BACKTRACE_LINE_LENGTH];
+        format_backtrace_line(i, &mStack[i], &symbols[i],
+                line, MAX_BACKTRACE_LINE_LENGTH);
+        ALOGD("%s%s", prefix, line);
+    }
+    free_backtrace_symbols(symbols, mCount);
+}
+#endif
+
 void CallStack::dump(int fd, int indent, const char* prefix) const {
     FdPrinter printer(fd, indent, prefix);
     print(printer);
